### PR TITLE
WIP: Ben/finetune from generator

### PR DESCRIPTION
**Pull request by lhz1029**

### CODE REVIEW

File: finetune/input_pipeline.py [hunk 1]
```diff
@@ -160,30 +160,19 @@ def _compute_class_counts(self, encoded_dataset):
...
-                raise ValueError( "Either none or all of Xs and Y and context should be callable, not a mixture")
-
-            dataset_encoded = lambda: itertools.chain.from_iterable(
-                map(lambda xyc: self.text_to_tokens_mask(*xyc), dataset())
-            )
+    def _dataset_with_targets(self, Xs, Y, train, context=None, update_hook=None, from_generator=False):
+        if from_generator:
+            def dataset():
+                for x in Xs():
+                    yield x.get("X"), x.get("Y"), x.get("context")
```
Comment (benleetownsend) [inline finetune/input_pipeline.py]: Casing inconsistencies on the Y

File: finetune/input_pipeline.py [hunk 2]
```diff
@@ -322,6 +320,101 @@ def internal_gen():
...
+                    "Generator input function does not have a length and no `config.dataset_size` is specified. "
+                    "You must set `config.dataset_size` explicitly."
+                )
+
+        self.config.val_size, self.config.val_interval = self.validation_settings(
+            n_examples=len(Xs) if not callable(Xs) else self.config.dataset_size,
+            batch_size=batch_size or self.config.batch_size,
+        )
+        self.config.dataset_size -= val_size
+        self._post_data_initialization(Y=None)
```
Comment (benleetownsend) [inline finetune/input_pipeline.py]: Sometimes we do have targets in this path.

File: finetune/input_pipeline.py [hunk 3]
```diff
@@ -322,6 +320,101 @@ def internal_gen():
...
+        self._skip_tqdm = val_size
+        dataset = self._make_dataset(Xs, None, train=True, update_hook=update_hook, from_generator=True, has_targets=has_targets)
+        val_dataset_unbatched = (
```
Comment (benleetownsend) [inline finetune/input_pipeline.py]: Factor together with the near duplicate code in get_train_input_fns

File: finetune/saver.py [hunk 4]
```diff
@@ -48,7 +48,13 @@ def __init__(
     def stop_if_no_metric_improvement_fn(self):
         if not self.keep_best_model:
             return False
-        eval_results = read_eval_metrics(self.estimator.eval_dir())
+        
+        eval_results = {
+            k: v for k, v in
+            read_eval_metrics(self.estimator.eval_dir()).items()
+            if "loss" in v # Filter out anything in eval that does not have a loss value
```
Comment (benleetownsend) [inline finetune/saver.py:55]: Does this need to be configurable, probably not but

File: finetune/target_models/masked_language_model.py [hunk 5]
```diff
@@ -208,7 +208,7 @@ def predict_top_k_report(self, input_text, k=5, context=None, forced_mask=None,
                                              for i in range(len(tokens))]) + 2*'\n'
 
         for i in sorted(predicted_tokens, key=lambda x: x['position']):
-            text_to_display += (f"{'<':>3}{mask_positions.index(i['position']):>2}>|{i['original_token_id']:15}|{i['prediction_ids']}\n")
+            text_to_display += ""#(f"{'<':>3}{mask_positions.index(i['position']):>2}>|{i['original_token_id']:15}|{i['prediction_ids']}\n")
```
Comment (benleetownsend) [inline finetune/target_models/masked_language_model.py]: oops was running on python 3.6.

File: finetune/base.py [hunk 6]
```diff
@@ -210,7 +221,12 @@ def finetune(self, Xs, Y=None, batch_size=None, context=None, update_hook=None):
                             name=task,
                         )
                     )
-            early_stopping_interval = sys.maxsize  # turn off early stopping for multi val
+            if None in val_input_fn:
+                saver_hook_interval = val_interval # has a default validation
+            else:
+                if self.config.keep_best_model:
+                    raise ValueError("Cannot use keep_best_model with MTL")
```
Comment (lhz1029) [inline finetune/base.py:237]: Is this just something that hasn't yet been implemented? Maybe NotImplementedError makes more sense?

File: finetune/input_pipeline.py (hunk 1 above)
Comment (lhz1029) [inline finetune/input_pipeline.py]: This is capital Y whereas it's lowercase y in dataset_without_targets

File: finetune/input_pipeline.py [hunk 7]
```diff
@@ -322,6 +320,101 @@ def internal_gen():
 
         return internal_gen()
 
+    def get_train_input_fns_from_generator(self, Xs, batch_size=None, val_size=None, update_hook=None):
+        assert callable(Xs)
+        self.epoch = 1
```
Comment (madisonmay) [inline finetune/input_pipeline.py]: What's with this business?

File: finetune/input_pipeline.py [hunk 8]
```diff
@@ -209,13 +198,19 @@ def _format_for_pipeline(data):
...
-        Xs_ = self._format_for_pipeline(Xs)
-        ys_ = self._format_for_pipeline(None)
-        context_ = self._format_for_pipeline(context)
-        forced_mask_ = self._format_for_pipeline(forced_mask)
+    def _dataset_without_targets(self, Xs, train, context=None, update_hook=None, forced_mask=None, from_generator=False):
+        if from_generator:
+            assert forced_mask is None, "Forced mask is not supported with generator inputs"
+            def Xs_gen():
+                for x in Xs():
+                    yield (x.get("X"), x.get("y"), x.get("context"), None)
```
Comment (lhz1029) [inline finetune/input_pipeline.py:204]: we could get into a strange situation if 'y' was actually provided for some reason

File: finetune/input_pipeline.py [hunk 9]
```diff
@@ -322,6 +320,101 @@ def internal_gen():
...
+                )
+
+        self.config.val_size, self.config.val_interval = self.validation_settings(
+            n_examples=len(Xs) if not callable(Xs) else self.config.dataset_size,
+            batch_size=batch_size or self.config.batch_size,
+        )
+        self.config.dataset_size -= val_size
+        self._post_data_initialization(Y=None)
+        self._skip_tqdm = val_size
+        dataset = self._make_dataset(Xs, None, train=True, update_hook=update_hook, from_generator=True, has_targets=has_targets)
```
Comment (madisonmay) [inline finetune/input_pipeline.py]: TODO: pass Ys along.

File: finetune/input_pipeline.py [hunk 10]
```diff
@@ -269,11 +264,14 @@ def validation_settings(self, n_examples, batch_size):
     def resampling(self, Xs, Y, context=None):
         return Xs, Y, context
 
-    def _make_dataset(self, Xs, Y, train=False, context=None, update_hook=None):
+    def _make_dataset(self, Xs, Y, train=False, context=None, update_hook=None, from_generator=False, has_targets=None):
         if Y is not None:
```
Comment (madisonmay) [inline finetune/input_pipeline.py]: Weird to check Y given that we have the has_targets setting

File: finetune/input_pipeline.py [hunk 11]
```diff
@@ -269,11 +264,14 @@ def validation_settings(self, n_examples, batch_size):
...
-            dataset = lambda: self._dataset_with_targets(Xs, Y, train=train, context=context, update_hook=update_hook)
+            has_targets = True
```
Comment (madisonmay) [inline finetune/input_pipeline.py]: Weird to check Y given that we're using the has_targets setting.  Is that just for the generator case?

I think long term we fix some of this code flow by keeping things bundled together until the last possible moment when we split them out?

File: finetune/target_models/masked_language_model.py (hunk 5 above)
Comment (madisonmay) [inline finetune/target_models/masked_language_model.py]: Assuming this is the typo you were speaking about

Comment (lhz1029) [inline finetune/target_models/masked_language_model.py]: is this intentional?

File: finetune/input_pipeline.py [hunk 12]
```diff
@@ -164,7 +164,6 @@ def _dataset_with_targets(self, Xs, Y, train, context=None, update_hook=None):
         if callable(Xs):
             raise ValueError('Cannot pass in generator for supervised dataset')
         else:
-            context = self._format_for_pipeline(context)
             dataset = lambda: zip(Xs, Y, context)
```
Comment (benleetownsend) [inline finetune/input_pipeline.py]: What if context is None?

File: finetune/input_pipeline.py (hunk 8 above)
Comment (benleetownsend) [reply finetune/input_pipeline.py:204]: Can we just replace this with a None then?

File: finetune/input_pipeline.py [hunk 13]
```diff
@@ -353,7 +346,8 @@ def get_train_input_fns(self, Xs, Y=None, batch_size=None, val_size=None, contex
         else:
             self._post_data_initialization(Y=None)
 
-        if callable(Xs) or Y is None:
+        if callable(Xs):
```
Comment (benleetownsend) [inline finetune/input_pipeline.py:350]: So if Y is none and not a callable (eg normal language model training) we now hit the next block where we try to shuffle None

File: finetune/input_pipeline.py [hunk 14]
```diff
@@ -210,12 +196,18 @@ def _format_for_pipeline(data):
...
     def _dataset_without_targets(self, Xs, train, context=None, update_hook=None, forced_mask=None):
-        Xs_ = self._format_for_pipeline(Xs)
-        ys_ = self._format_for_pipeline(None)
-        context_ = self._format_for_pipeline(context)
-        forced_mask_ = self._format_for_pipeline(forced_mask)
+        if callable(Xs):
+            assert forced_mask is None, "Forced mask is not supported with generator inputs"
+            def Xs_gen():
+                for x in Xs():
+                    yield (x.get("X"), x.get("y"), x.get("context"), None)
```
Comment (benleetownsend) [inline finetune/input_pipeline.py:204]: This does break generate_text and the cached predict that is on this branch, but i'm not sure if thats okay?

**Pull request merged**